File: app/assets/javascripts/game/Rank.js
define(function () {
  "use strict";

  var two = { value: 2, char: "2", name: "Two" };
  var three = { value: 3, char: "3", name: "Three" };
  var four = { value: 4, char: "4", name: "Four" };
  var five = { value: 5, char: "5", name: "Five" };
  var six = { value: 6, char: "6", name: "Six" };
  var seven = { value: 7, char: "7", name: "Seven" };
  var eight = { value: 8, char: "8", name: "Eight" };
  var nine = { value: 9, char: "9", name: "Nine" };
  var ten = { value: 10, char: "X", name: "Ten" };
  var jack = { value: 11, char: "J", name: "Jack" };
  var queen = { value: 12, char: "Q", name: "Queen" };
  var king = { value: 13, char: "K", name: "King" };
  var ace = { value: 14, char: "A", name: "Ace" };
  var unknown = { value: 0, char: "?", name: "Unknown" };

  var Rank = {
    two: two,
    three: three,
    four: four,
    five: five,
    six: six,
    seven: seven,
    eight: eight,
    nine: nine,
    ten: ten,
    jack: jack,
    queen: queen,
    king: king,
    ace: ace,
    unknown: unknown,

    all: [two, three, four, five, six, seven, eight, nine, ten, jack, queen, king, ace],
    fromChar: function(c) {
      switch(c) {
        case '2': return two;
        case '3': return three;
        case '4': return four;
        case '5': return five;
        case '6': return six;
        case '7': return seven;
        case '8': return eight;
        case '9': return nine;
        case 'X': return ten;
        case 'J': return jack;
        case 'Q': return queen;
        case 'K': return king;
        case 'A': return ace;
        case '?': return unknown;
        default: throw "Invalid rank [" + c + "].";
      }
    }
  };

  return Rank;
});

File: shared/src/main/scala/models/card/Rank.scala
package models.card

import enumeratum.values._

sealed abstract class Rank(val value: Char) extends CharEnumEntry {
  def index: Int
  def name: String
  lazy val previous = Rank.allByValue(index - 1)
  lazy val next = Rank.allByValue(index + 1)
  val locs: Seq[(Double, Double)] = Nil
}

object Rank extends CharEnum[Rank] with CharCirceEnum[Rank] {
  case object Two extends Rank('2') {
    override val index = 2
    override val name = "Two"
    override lazy val previous = Ace
    override val locs = Seq((0.5, 0.3), (0.5, 0.7))
  }

  case object Three extends Rank('3') {
    override val index = 3
    override val name = "Three"
    override val locs = Seq((0.2, 0.7), (0.5, 0.5), (0.8, 0.3))
  }

  case object Four extends Rank('4') {
    override val index = 4
    override val name = "Four"
    override val locs = Seq((0.2, 0.5), (0.5, 0.3), (0.5, 0.7), (0.8, 0.5))
  }

  case object Five extends Rank('5') {
    override val index = 5
    override val name = "Five"
    override val locs = Seq((0.2, 0.3), (0.8, 0.3), (0.5, 0.5), (0.2, 0.7), (0.8, 0.7))
  }

  case object Six extends Rank('6') {
    override val index = 6
    override val name = "Six"
    override val locs = Seq((0.2, 0.4), (0.2, 0.6), (0.5, 0.3), (0.5, 0.7), (0.8, 0.4), (0.8, 0.6))
  }

  case object Seven extends Rank('7') {
    override val index = 7
    override val name = "Seven"
    override val locs = Seq((0.2, 0.4), (0.2, 0.6), (0.5, 0.3), (0.5, 0.5), (0.5, 0.7), (0.8, 0.4), (0.8, 0.6))
  }

  case object Eight extends Rank('8') {
    override val index = 8
    override val name = "Eight"
    override val locs = Seq((0.2, 0.3), (0.2, 0.5), (0.2, 0.7), (0.5, 0.4), (0.5, 0.6), (0.8, 0.3), (0.8, 0.5), (0.8, 0.7))
  }

  case object Nine extends Rank('9') {
    override val index = 9
    override val name = "Nine"
    override val locs = Seq((0.2, 0.4), (0.2, 0.6), (0.2, 0.8), (0.5, 0.3), (0.5, 0.5), (0.5, 0.7), (0.8, 0.2), (0.8, 0.4), (0.8, 0.6))
  }

  case object Ten extends Rank('X') {
    override val index = 10
    override val name = "Ten"
    override val locs = Seq((0.2, 0.3), (0.2, 0.5), (0.2, 0.7), (0.5, 0.2), (0.5, 0.4), (0.5, 0.6), (0.5, 0.8), (0.8, 0.3), (0.8, 0.5), (0.8, 0.7))
  }

  case object Jack extends Rank('J') {
    override val index = 11
    override val name = "Jack"
  }

  case object Queen extends Rank('Q') {
    override val index = 12
    override val name = "Queen"
  }

  case object King extends Rank('K') {
    override val index = 13
    override val name = "King"
  }

  case object Ace extends Rank('A') {
    override val index = 14
    override val name = "Ace"
    override lazy val next = Two
  }

  case object Unknown extends Rank('?') {
    override val index = 0
    override val name = "Unknown"
    override lazy val previous = Unknown
    override lazy val next = Unknown
  }

  override val values = findValues
  val all = values.filterNot(_ == Unknown)
  val allByValue = all.map(r => r.index -> r).toMap + (1 -> Ace)
}
